mozfldp/server.py

```diff
@@ -24,74 +19,120 @@ class ServerFacade:
-        pass
-
-    def update_classifier(self):
-        # TODO: this is where the magic happens
-        #
-        # I think roughly, this should be applying a sample of client
-        # data through a training loop.
-        #
-        # I believe you want to take the code from:
-        # https://github.com/mozilla/CANOSP-2019/blob/master/simulation_util.py#L98
-        # to
-        # https://github.com/mozilla/CANOSP-2019/blob/master/simulation_util.py#L149
-        # and update the parameters of the SGDClassifier.
-        #
-        # The samples `S` in simulation_util.py would be equivalent to self._buffer
-        # in this class
-        pass
-
-    def classify(self, some_input_json):
         """
-        Apply the classifer to some sample input
+        Accepts new weights from a client and stores them on the server side for averaging
+        
+        Args: 
+            client_json: a json object containing coefs, intercepts, and num_samples
+        """
+        client_json = json.loads(client_json)
+        self._client_coefs.append(client_json["coefs"])
+        self._client_intercepts.append(client_json["intercept"])
+        self._num_samples.append(client_json["num_samples"])
+
+    def compute_new_weights(self):
+        """ 
+        Applies the federated averaging on the stored client weights for this round
+        and return the new weights
         """
-        reshaped_X_test = None
-        reshaped_Y_test = None
-        score = self._classifier.score(reshaped_X_test, reshaped_Y_test)
-        return score
 
+        new_coefs = np.zeros(self._coef.shape, dtype=np.float64, order="C")
+        new_intercept = np.zeros(self._intercept.shape, dtype=np.float64, order="C")
 
-server = ServerFacade()
+        for index, (client_coef, client_intercept) in enumerate(
+            zip(self._client_coefs, self._client_intercepts)
+        ):
+            n_k = self._num_samples[index]
+            added_coef = [
+                np.array(value) * (n_k) / sum(self._num_samples)
+                for value in client_coef
+            ]
+            added_intercept = [
+                np.array(value) * (n_k) / sum(self._num_samples)
+                for value in client_intercept
+            ]
 
-# TODO: you'll need to add server routes for `update_classifier` and
-# `classify` to allow them to be invoked over the web.
+            new_coefs = np.add(new_coefs, added_coef)
+            new_intercept = np.add(new_intercept, added_intercept)
 
+        # update the server weights to newly calculated weights
+        print("Updated Weights: ", new_coefs, new_intercept)
 
-@app.route("/api/v1/client_update/<string:client_id>", methods=["POST"])
-def client_update(client_id):
-    payload = request.get_json()
+        self._coef = new_coefs
+        self._intercept = new_intercept
+
+        # reset the state for the next round of learning
+        self._client_coefs = None
+        self._client_intercepts = None
```

Comment:
I think these should be set to empty lists. Since the same reset logic is used in the constructor, maybe wrap these statements in a "reset" function that can be called in both places.

Reply:
Done